dialogue-core/src/main/java/com/palantir/dialogue/core/QueuedChannel.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.common.util.concurrent.SettableFuture;
+import com.palantir.conjure.java.api.errors.QosException;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import java.util.Deque;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.Executor;
+import org.immutables.value.Value;
+
+final class QueuedChannel implements Channel {
+
+    private static final Executor DIRECT = MoreExecutors.directExecutor();
+    private final Deque<CallComponents> queuedCalls = new ConcurrentLinkedDeque<>();
+    private final LimitedChannel delegate;
+
+    QueuedChannel(LimitedChannel delegate) {
+        this.delegate = delegate;
+    }
+
+    /**
+     * Enqueue the call and try to schedule as many queued tasks as possible.
+     */
+    @Override
+    public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+        CallComponents components = ImmutableCallComponents.of(endpoint, request, SettableFuture.create());
+
+        if (!queuedCalls.offer(components)) {
+            throw QosException.unavailable();
+        }
+
+        schedule();
```

Comment:
it's a little weird to me that we 'steal' this thread to possibly cause the scheduling/processing of other in-flight calls; maybe we should just eat the cost of having a thread around to do the processing of the queue and have it poll/wait for work to show up?

Reply:
yeah I'm +1 to keeping a thread around for scheduling. I've been trying to figure out the right way to signal to that thread that it should run. Don't think I want to be in the business fo `wait`/`notify`, but that's basically the behavior you want (scheduling thread `wait`s until a request is enqueued or completed, and then should be `notified`).

Reply:
the issues is that `handle(call)` may not be able to execute (the LimitedChannel can return `Optional.empty()`)